minpy/tape.py

```diff
@@ -0,0 +1,209 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# pylint: disable=logging-format-interpolation
+"""Tape for recording gradient calculation."""
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import with_statement
+import contextlib
+import collections
+
+import numpy
+
+from . import array
+from . import array_variants
+from .utils import log
+
+# pylint: disable=invalid-name
+_logger = log.get_logger(__name__)
+# pylint: enable=invalid-name
+
+GradRecord = collections.namedtuple(
+    'GradRecord', ['grad_func', 'result', 'primitive_type', 'owner'])
+
+
+class Tape(object):
+    """Records gradient calculation."""
+    global_tape = None
+
+    def __init__(self):
+        self._grads = {}
+        self._array_grad_records = {}
+
+    def add_partial_derivative(self, grad_func, owner, result, primitive_type):
+        """Add partial derivative.
+
+        Parameters
+        ----------
+        grad_func
+            Function for calculating gradient.
+        owner
+            Owners of the gradient. Usually it is just one array. But a list is also allowed.
+        result
+            Result of previous forward execution.
+        primitive_type
+            Primitive type indicating location of execution.
+        """
+
+        def add_owner(single_owner):
```

Comment:
Does the argument name single_owner sounds a little ambiguous? I was wondering if the "if" logic of determining the type of "single_owner" could be moved into its parent function?